src/hotspot/share/gc/g1/g1RegionPinCache.cpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "gc/g1/g1RegionPinCache.inline.hpp"
+#include "gc/g1/g1CollectedHeap.inline.hpp"
+#include "gc/g1/heapRegion.inline.hpp"
+
+G1RegionPinCache::~G1RegionPinCache() {
+  flush();
+}
```

Comment:
This file is almost empty; I wonder if this method can be moved to hpp or inline.hpp.

Reply:
Moving this to the .hpp or .inline.hpp would require to include the .inline.hpp into g1ThreadLocalData.hpp, which we do not allow.
The alternative would be creating a g1ThreadLocalData.inline.hpp and putting it there, but that would gain nothing.
I'll try.

Reply:
Doing the flush in the destructor is too late.  It needs to be done in G1BarrierSet::on_thread_detach.
A thread might be detached but not yet deleted for SMR reasons.  If a GC occurs then, the pending
cached pin manipulation won't be found by the thread walk because the thread is not in the current
threads list.

That of course solves the problem of where to put it. :)